File: career/frontend/types/badge.ts
// Badge Types - matches badge-service (port 8012) schemas

export type BadgeCategory =
  | 'skill'          // 스킬 마스터리
  | 'achievement'    // 성취
  | 'course'         // 강의 수료
  | 'project'        // 프로젝트 완료
  | 'certification'  // 자격증
  | 'competition'    // 공모전
  | 'community';     // 커뮤니티 기여

export type BadgeLevel = 'bronze' | 'silver' | 'gold' | 'platinum';

// Badge
export interface BadgeBase {
  badge_cd: string;
  badge_nm: string;
  description?: string;
  category: BadgeCategory;
  level: BadgeLevel;
  icon_url?: string;
  criteria?: string;
  points: number;
  related_skill_cd?: string;
  is_active: boolean;
}

export interface BadgeCreate extends BadgeBase {}

export interface BadgeResponse extends BadgeBase {
  badge_id: string;
  created_at: string;
}

// Student Badge
export interface StudentBadgeResponse {
  student_badge_id: string;
  student_id: string;
  badge_id: string;
  badge: BadgeResponse;
  earned_at: string;
  issued_by?: string;
  evidence_url?: string;
  verification_hash?: string;
  is_public: boolean;
}

export interface BadgeAwardRequest {
  student_id: string;
  badge_cd: string;
  issued_by?: string;
  evidence_url?: string;
  is_public?: boolean;
}

// Skill Credential
export interface SkillCredential {
  skill_cd: string;
  skill_nm: string;
  level: number;
  verified: boolean;
  verification_source?: string;
  verified_at?: string;
  related_badges: string[];
  evidence: string[];
}

// Credential Response (for getStudentCredentials API)
export interface CredentialResponse {
  credential_id: string;
  student_id: string;
  credential_type: 'certification' | 'license' | 'award' | 'completion';
  credential_nm: string;
  issuing_org: string;
  issue_date: string;
  expiry_date?: string;
  credential_number?: string;
  verification_url?: string;
  verified: boolean;
  related_skill_cd?: string;
  created_at: string;
}

// Skill Passport
export interface SkillPassportResponse {
  passport_id: string;
  student_id: string;
  student_name?: string;
  issued_at: string;
  updated_at: string;

  // Summary
  total_skills: number;
  verified_skills: number;
  total_badges: number;
  total_points: number;

  // Credentials
  skills: SkillCredential[];
  badges: StudentBadgeResponse[];

  // Verification
  passport_hash?: string;
  is_valid: boolean;
  share_url?: string;
}

// Passport Share
export interface PassportShareRequest {
  student_id: string;
  include_skills?: boolean;
  include_badges?: boolean;
  expiry_days?: number;  // 1-365
  recipient_email?: string;
}

export interface PassportShareResponse {
  share_id: string;
  share_url: string;
  expires_at: string;
  access_count: number;
  created_at: string;
}

// Helper
export const BADGE_LEVEL_COLORS: Record<BadgeLevel, string> = {
  bronze: '#cd7f32',    // bronze
  silver: '#c0c0c0',    // silver
  gold: '#ffd700',      // gold
  platinum: '#e5e4e2',  // platinum
};

export const BADGE_LEVEL_LABELS: Record<BadgeLevel, string> = {
  bronze: '브론즈',
  silver: '실버',
  gold: '골드',
  platinum: '플래티넘',
};

export const BADGE_CATEGORY_LABELS: Record<BadgeCategory, string> = {
  skill: '스킬 마스터리',
  achievement: '성취',
  course: '강의 수료',
  project: '프로젝트',
  certification: '자격증',
  competition: '공모전',
  community: '커뮤니티',
};

export const BADGE_CATEGORY_ICONS: Record<BadgeCategory, string> = {
  skill: '⚡',
  achievement: '🏆',
  course: '📚',
  project: '💻',
  certification: '📜',
  competition: '🥇',
  community: '🤝',
};

File: career/services/badge-service/app/schemas/badge.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import date, datetime
from uuid import UUID
from enum import Enum


class BadgeCategory(str, Enum):
    SKILL = "skill"             # 스킬 마스터리
    ACHIEVEMENT = "achievement" # 성취
    COURSE = "course"           # 강의 수료
    PROJECT = "project"         # 프로젝트 완료
    CERTIFICATION = "certification"  # 자격증
    COMPETITION = "competition" # 공모전
    COMMUNITY = "community"     # 커뮤니티 기여
    COMPETENCY = "competency"   # 역량
    MILESTONE = "milestone"     # 마일스톤


class BadgeLevel(str, Enum):
    BRONZE = "bronze"
    SILVER = "silver"
    GOLD = "gold"
    PLATINUM = "platinum"


class BadgeBase(BaseModel):
    badge_cd: str
    badge_nm: str
    description: Optional[str] = None
    category: BadgeCategory
    level: BadgeLevel = BadgeLevel.BRONZE
    icon_url: Optional[str] = None
    criteria: Optional[Dict[str, Any]] = None
    points: int = 10
    related_skill_cd: Optional[str] = None
    is_active: bool = True


class BadgeCreate(BadgeBase):
    pass


class BadgeResponse(BadgeBase):
    badge_id: UUID
    created_at: datetime

    class Config:
        from_attributes = True


class StudentBadgeResponse(BaseModel):
    student_badge_id: UUID
    student_id: str
    badge_id: UUID
    badge: BadgeResponse
    earned_at: datetime
    issued_by: Optional[str] = None
    evidence_url: Optional[str] = None
    verification_hash: Optional[str] = None
    is_public: bool = True

    class Config:
        from_attributes = True


class BadgeAwardRequest(BaseModel):
    student_id: str
    badge_cd: str
    issued_by: Optional[str] = None
    evidence_url: Optional[str] = None
    is_public: bool = True


class SkillCredential(BaseModel):
    skill_cd: str
    skill_nm: str
    level: int
    verified: bool = False
    verification_source: Optional[str] = None
    verified_at: Optional[datetime] = None
    related_badges: List[str] = Field(default_factory=list)
    evidence: List[str] = Field(default_factory=list)


class SkillPassportResponse(BaseModel):
    passport_id: UUID
    student_id: str
    student_name: Optional[str] = None
    issued_at: datetime
    updated_at: datetime

    # Summary
    total_skills: int = 0
    verified_skills: int = 0
    total_badges: int = 0
    total_points: int = 0

    # Credentials
    skills: List[SkillCredential] = Field(default_factory=list)
    badges: List[StudentBadgeResponse] = Field(default_factory=list)

    # Verification
    passport_hash: Optional[str] = None
    is_valid: bool = True
    share_url: Optional[str] = None

    class Config:
        from_attributes = True


class PassportShareRequest(BaseModel):
    student_id: str
    include_skills: bool = True
    include_badges: bool = True
    expiry_days: int = Field(default=30, ge=1, le=365)
    recipient_email: Optional[str] = None


class PassportShareResponse(BaseModel):
    share_id: UUID
    share_url: str
    expires_at: datetime
    access_count: int = 0
    created_at: datetime


class CredentialType(str, Enum):
    CERTIFICATION = "certification"  # 자격증
    CERTIFICATE = "certificate"      # 자격증 (alternate)
    LICENSE = "license"              # 면허
    AWARD = "award"                  # 수상
    COMPLETION = "completion"        # 수료
    PUBLICATION = "publication"      # 논문/출판물
    SCHOLARSHIP = "scholarship"      # 장학금
    PATENT = "patent"                # 특허
    LANGUAGE = "language"            # 어학


class CredentialResponse(BaseModel):
    credential_id: UUID
    student_id: str
    credential_type: CredentialType
    credential_nm: str
    issuing_org: str
    issue_date: date
    expiry_date: Optional[date] = None
    credential_number: Optional[str] = None
    verification_url: Optional[str] = None
    verified: bool = False
    related_skill_cd: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True
